core/src/main/java/org/apache/iceberg/ScanProperties.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+public class ScanProperties {
```

Comment:
I am thinking about the best way to support this. I agree about having a Spark read option but I wonder whether we should have scan options or some sort of `useSnapshots(ids)` method in `TableScan`. We can only support that method in `ALL_*` metadata tables and throw an exception otherwise.

Thoughts, @rdblue @szehon-ho @RussellSpitzer?

Reply:
This isn't terrible, but I would much rather expose this in some other way. I don't think adding it to `TableScan` makes sense because it is very specific to this feature and would cause confusion in other cases. A data scan, for example, can't scan more than one snapshot.

What about exposing a `reference_snapshot_id` column (or with an _ for a metadata column)? If we did that, we could accept a filter from Spark, like `reference_snapshot_id IN (1, 2, 3, 4)` and special case the pushdown. We'd need to ensure that we give a consistent value for that column, which could be the latest snapshot that has the manifest.

Reply:
@rdblue yea , I considered that.  My thought at the time is that same manifest may have different reference_snapshot_id .  Currently I guess we don't do de-duplication in all-manifest, but this would enforce that we return one manifest per reference_snapshot_id (ie, for an append-only table, the first manifest would appear as many times as snapshots.).  Also seems like this would require a bit complex code like adding a ProjectionEvaluator for snapshot filters.  Not totally against it, and maybe this is a useful feature even outside, but seems a bit harder to do in my view.

Reply:
> A data scan, for example, can't scan more than one snapshot.

It is true but `ALL_*` tables, on the other hand, don't support `useSnapshot`, `asOfTime` and incremental features.

> What about exposing a reference_snapshot_id column

I'd be open for that too. It will be the snapshot ID of the manifest list we read these manifests from, right? Adding a new column is fairly easy. How can we do the filtering? We need to filter `Snapshot`s and we have a generic filter `Expression` that can reference any columns. Is there an easy way to perform the filtering?

Reply:
Yeah, I think to perform the filtering we just need to go through all the snapshots and evaluate the expression on each one.

Reply:
Essentially a custom expression visitor applied on `Snapshot` object?

Reply:
@aokolnychyi, that, or we could just pattern match and reject anything but IN predicates. As long as we're the ones controlling what filters we use, we have some flexibility.